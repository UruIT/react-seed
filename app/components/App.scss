@import '../styles/box-shadow.scss';

.shadow {
	@include box-shadow(1px, 1px, 4px,#4c4a4a);
}

.container {
	width: 420px;
	padding: 10px;
	margin: 32px auto;
	text-align: center;
}

.button {
	border: 2px solid #000;
	background-color: #fbfb09;
	cursor: pointer;
	font-weight: 600;
	font-size: 2rem;
	width: 200px;
	margin: 2rem;
}
