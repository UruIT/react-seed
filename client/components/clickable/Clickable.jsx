import React from 'react';
import PropTypes from 'prop-types';
import noop from 'utils/noop';
import styles from './clickable.scss';

const Clickable = ({ content, onClick }) => (
	<div className={styles.container} onClick={onClick}>
		{content}
	</div>
);

Clickable.propTypes = {
	content: PropTypes.node,
	onClick: PropTypes.func
};

Clickable.defaultProps = {
	content: '...',
	onClick: noop
};

export default Clickable;
