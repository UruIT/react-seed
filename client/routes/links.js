const index = '/';

const about = '/about';

const chucknorris = 'https://api.chucknorris.io/jokes/random';

export default {
	index,
	about,
	chucknorris
};
