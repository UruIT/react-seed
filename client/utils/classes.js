function classes(...classnames) {
	return classnames.filter(classname => !!classname).join(' ');
}

export default classes;
