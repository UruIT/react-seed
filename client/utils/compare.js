const compare = (a, b) => Math.sign(a.index - b.index);

export { compare };
