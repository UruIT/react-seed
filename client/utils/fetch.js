import 'es6-promise/auto';
import 'whatwg-fetch';

const _errorHandle = response => {
	if (response.ok) {
		return response;
	}

	let error = new Error(response.statusText);
	error.status = response.status;
	error.response = response;

	throw error;
};

const _parseJson = response => response.json();

export const getJson = url => fetch(url).then(_errorHandle).then(_parseJson);

export default { getJson };
