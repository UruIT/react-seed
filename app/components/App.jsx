import React from 'react';
import MyButton from './MyButton';
import styles from './App.scss';
import { getJson } from '../utils/fetch';

export default class App extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			joke: ''
		};
	}
	componentDidMount() {
		getJson('https://api.chucknorris.io/jokes/random')
			.then(res => this.setState({
				joke: res.value
			}));
	}
	render() {
		return (
			<div>
				<button className={styles.button} onClick={this.test}>Test</button>
				<MyButton>{ this.state.joke }</MyButton>
			</div>
		);
	}

	test = () => {
		console.error('test')
	}
}
