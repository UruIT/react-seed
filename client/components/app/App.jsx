import React from 'react';
import { getJson } from 'utils/fetch';
import links from '../../routes/links';
import Counter from '../counter/Counter';
import MyButton from '../button/Button';
import styles from './app.scss';
import classes from 'utils/classes';

class App extends React.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
		this.state = {
			joke: ''
		};
	}

	handleClick() {
		return getJson(links.chucknorris).then(res =>
			this.setState({
				joke: res.value
			})
		);
	}

	render() {
		const { joke } = this.state;

		return (
			<div className={styles.container}>
				<Counter />
				<button className={classes(styles.button, styles.shadow)} onClick={this.handleClick}>
					Test
				</button>
				{!!joke && <MyButton text={joke} />}
			</div>
		);
	}
}

export default App;
