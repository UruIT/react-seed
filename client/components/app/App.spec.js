import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow, mount } from 'enzyme';

import { getJson } from 'utils/fetch';
import Clickable from '../clickable';
import links from '../../routes/links';
import App from './App';

jest.mock('utils/fetch', () => ({
	getJson: jest.fn(url => new Promise((resolve, reject) => (url ? resolve({ value: 'another joke' }) : reject())))
}));

describe('<App/>', () => {
	it('snapshot', () => {
		const renderer = new ShallowRenderer();
		const tree = renderer.render(<App />);
		expect(tree).toMatchSnapshot();
	});
});

describe('App - button onClick', () => {
	const app = shallow(<App />);

	beforeEach(() => {
		getJson.mockImplementation(() => Promise.resolve({ value: 'another joke' }));
		app.find('button').simulate('click');
	});

	it('should call ChuckNorris api', () => {
		expect(getJson).toBeCalledWith(links.chucknorris);
	});

	it('should keep previous jokes', () => {
		expect(app.state().jokes.length).toBe(2);
	});

	it('should update <Clickable> children', () => {
		return app
			.instance()
			.handleClick()
			.then(() =>
				expect(
					app
						.find(Clickable)
						.first()
						.props().content
				).toEqual('another joke')
			);
	});
});

describe('App - mounting', () => {
	describe('server request succeeded', () => {
		const sample = [{ id: 1 }, { id: 2 }];
		getJson.mockImplementation(() => Promise.resolve(sample));
		const tree = mount(<App />);

		it('should call getJson', () => {
			expect(getJson).toBeCalledWith(links.api.sample);
		});

		it('should update state.sample', () => {
			expect(tree.instance().state).toMatchObject({ sample });
		});
	});

	describe('server request error', () => {
		getJson.mockImplementation(() => Promise.reject({ code: 500 }));
		const app = mount(<App />);

		it('should update state.error', () => {
			expect(app.instance().state).toMatchObject({ error: { code: 500 } });
		});

		it('should not have any joke', () => {
			expect(app.state().jokes).toEqual([]);
		});
	});
});
