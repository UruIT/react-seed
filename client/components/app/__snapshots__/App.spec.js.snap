// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> snapshot 1`] = `
<div
  className="container"
>
  <Counter />
  <button
    className="button shadow"
    onClick={[Function]}
  >
    Test
  </button>
</div>
`;
